### ISSUE TITLE
dropout enabled during sampling!

### ISSUE THREAD

**Issue by mrmartin** [jcjohnson/torch-rnn#31]
Going through the code, I found that dropout layers aren't disabled during sampling. This certainly isn't expected behaviour.

After training, I manually set all the dropout layers p variable to 0, and the probability distributions for a given initiation sequence stopped changing. Without doing this, I'd get different results across runs.

This presumably also means that dropout isn't being disabled during validation.

Can anyone confirm that this is a bug? If it is, I can do a fix in LanguageModel.lua.

**Comment by jcjohnson** [comment]
Setting the model to evaluate mode disables dropout, which is done both during validation and sampling:

https://github.com/jcjohnson/torch-rnn/blob/master/train.lua#L189
https://github.com/jcjohnson/torch-rnn/blob/master/sample.lua#L39

Getting different outputs across different runs when sampling is expected, since by default we are *sampling* from the probability distribution at each time step rather than just taking the argmax; there is randomness involved.

**Comment by jcjohnson** [comment]
Also note that simply setting the dropout `p` variable to zero will give incorrect results, since `nn.Dropout` uses inverted dropout that requires scaling at test-time:

https://github.com/torch/nn/blob/master/Dropout.lua#L49

**Comment by mrmartin** [comment]
Thanks for the quick reply. 
I added printing of the probability distribution into the LM:sample() function of LanguageModel.lua:
```lua
probs:div(torch.sum(probs))
print(probs)
next_char = torch.multinomial(probs, 1):view(1, 1)
```
 When I sample a single character using sample.lua, like this:
`th sample.lua -length 9 -start_text "How are " -checkpoint cv/Sentinel140_dropout_0.1_3_layers_256_rnn_33060.t7`, I get a slightly different distribution for each run of the code. This is not influenced by sampling. (I also get a different character each time, as expected.)
When I do the same thing with a net trained without dropout, the distribution is the same across runs.
When I set the dropout `p` value to 0 manually, the results might not be correct, but they look good, and they're consistent. 
And yes, the `model:evaluate()` is being called. Could this be a more fundamental flaw? Or am I doing something wrong?

**Comment by jcjohnson** [comment]
Hmm, that's weird. I'll take another look.

**Comment by FBMachine** [comment]
The parent class of LanguageModel is nn.Module, which only sets self.train when calling evaluate()/train(). Since LanguageModel is essentially a container, you need to forward the call to self.net in order for it to propagate to the actual modules of the net.

Note that you don't need to do this for model:cuda/cl() since Module:type() does a recursive walk through all members in the "self" table for you.

**Comment by jcjohnson** [comment]
Yep, that's the bug. Good catch!

**Comment by mrmartin** [comment]
I can confirm that! 
By replacing `model:evaluate()` by `model.net:evaluate()` in `sample.lua`, the problem is resolved.
This should also be done in `train.lua` and `eval.lua`, to fix the issue during validation.

**Comment by jcjohnson** [comment]
I implemented the fix suggested by @FBMachine in https://github.com/jcjohnson/torch-rnn/commit/b77a8b70435a901a1e85f86027a7c971ba76a518. Thanks for catching this!
